File: 33.MonotonicStackAlgo.java
/* ----------------------------------------------------------------------------  */
/*   ( The Authentic JS/JAVA CodeBuff )
 ___ _                      _              _ 
 | _ ) |_  __ _ _ _ __ _ __| |_ __ ____ _ (_)
 | _ \ ' \/ _` | '_/ _` / _` \ V  V / _` || |
 |___/_||_\__,_|_| \__,_\__,_|\_/\_/\__,_|/ |
                                        |__/ 
 */
/* --------------------------------------------------------------------------   */
/*    Youtube: https://youtube.com/@code-with-Bharadwaj                        */
/*    Github : https://github.com/Manu577228                                  */
/*    Portfolio : https://manu-bharadwaj-portfolio.vercel.app/portfolio      */
/* -----------------------------------------------------------------------  */

import java.io.*;
import java.util.*;

public class MonotonicStackNGE {

    // Function to find Next Greater Element
    public static int[] nextGreaterElement(int[] arr) {
        int n = arr.length;                // 1. Size of the array
        int[] res = new int[n];            // 2. Result array
        Arrays.fill(res, -1);              // 3. Initialize all values to -1
        Deque<Integer> st = new ArrayDeque<>(); // 4. Stack to hold indexes

        for (int i = 0; i < n; i++) {      // 5. Traverse array
            while (!st.isEmpty() && arr[i] > arr[st.peek()]) {
                // 6. If current element > element at top index of stack
                int idx = st.pop();        // 7. Pop index
                res[idx] = arr[i];         // 8. Set result for that index
            }
            st.push(i);                    // 9. Push current index
        }
        return res;                        // 10. Return result array
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        PrintWriter out = new PrintWriter(System.out);

        // Example input
        int[] arr = {2, 1, 5, 6, 2, 3};
        int[] output = nextGreaterElement(arr);

        out.println("Input : " + Arrays.toString(arr));
        out.println("Output: " + Arrays.toString(output));
        out.flush();
    }
}

File: 33.MonotonicStackAlgo.py
# A Monotonic Stack is a stack-based algorithm that maintains elements in either increasing or 
# decreasing order while processing. It is widely used to solve Next Greater Element, 
# Stock Span, Trapping Rain Water, and similar problems efficiently.

# Explanation

# Normally, finding the "next greater" or "next smaller" 
# element for each item in an array takes O(n²) if we check every element.

# But using a Monotonic Stack, we can do this in O(n) time.

# The stack keeps elements in sorted order (monotonic).

# Monotonic Increasing Stack → smallest at the bottom, largest at the top.

# Monotonic Decreasing Stack → largest at the bottom, smallest at the top.

# While traversing the array, we maintain this property by popping 
# elements that violate the order and pushing the current element.

# 3. Example Problem:

# Find the Next Greater Element (NGE) for each element in the array.
# Input: [2, 1, 5, 6, 2, 3]
# Output: [5, 5, 6, -1, 3, -1]

def next_greater_element(arr):
    n = len(arr)
    res = [-1] * n
    st = []

    for i in range(n):
        while st and arr[i] > arr[st[-1]]:
            idx = st.pop()
            res[idx] = arr[i]
        st.append(i)

    return res

arr = [2, 1, 5, 6, 2, 3]
output = next_greater_element(arr)

print("Input :", arr)
print("Output:", output)
